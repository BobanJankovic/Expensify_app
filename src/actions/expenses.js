import uuid from "uuid";
import database from '../firebase/firebase';

/*
expense: {
          id:expect.any(String),
          description: '',
          note: '',
          amount: 0, 
          createdAt: 0
        }

zbog parametra expense u expenses.test.js kada proveravamo toEqual ocekujemo expense objekat sa svime gore navedenim
*/ 
export const addExpense = (expense) => {
  return {
    type: "ADD_EXPENSE",
    expense
  }
};
export const startAddExpense = ( {description = '', note = '', amount = 0, createdAt=0 } = {}) => { 
  //funkcija umesto objekta vraca funkciju(koja radi 2 stvari
  // 1)pusuje nas expense objekat u firebase, a onda ako je promise uspesan dispacuje i salje u reducer )
  return (dispatch) => {
   const expense = {description, note, amount, createdAt };
   return database.ref('expenses').push(expense).then((ref)=>{
     dispatch(addExpense( {id: ref.key, description, note, amount, createdAt }));
     //add expense vrati objekat koji u sebi ima type:add_expense i objekat koji se zove expense 
     //i ima vredonsti parametara iznad
   });
   };
  };



export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});


