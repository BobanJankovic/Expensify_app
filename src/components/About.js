import React from 'react';
export const About = () => (
    <div>
    This is from  About
    </div>
);

 