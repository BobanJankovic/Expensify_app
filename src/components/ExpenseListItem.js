import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import uuid from "uuid";
import { Link } from 'react-router-dom'


//connect redux component to the redux store
const ExpenseListItem= (props)=> (
<div>
  <div>
    {props.expenses.length===0 && <p>Please add an expense</p>}
     {  props.expenses.map((expense) => {
       
          return (
            
            <div key={expense.id}>
              <p><Link to={`/edit/${expense.id}`} >{expense.description}</Link>:{expense.amount} created at : {expense.createdAt}</p>
             
            </div>
          );   
        })
      }
    </div>
    
    
  </div>

);

const mapStateToProps = (state)=> {
  return {
    //state.expenses je arej objekate, state.filters je objekat sa key-value parovima
    //expenses je filtrirani arej objekaata prema state.filters
    expenses:selectExpenses(state.expenses,state.filters)
    
  };
};
export default connect(mapStateToProps)(ExpenseListItem);



