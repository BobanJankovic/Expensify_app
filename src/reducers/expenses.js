// Expenses Reducer

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
      
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if(expense.id===action.id){
           // console.log(action.updates);
           //console.log(expense);
           //uzima nas stari objekat expense ceo i overriduje njegov jedini key amount iz updates objekta
         return {...expense, ...action.updates}
          
        }else {
          expense;
        }
      });
    default:
      return state;
  }
};
