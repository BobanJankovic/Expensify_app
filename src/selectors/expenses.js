//expenses-arej sa objektima
//kada ga exportujem default u importu mogu da napisem ime koje god hocu.

export default (expenses, {text, sortBy, startDate, endDate})=> {
  return expenses.filter((expense)=>{
    const startDateMatch= typeof startDate !== "number" || expense.createdAt >= startDate;
    const endDateMatch= typeof startDate !== "number" || expense.createdAt <= endDate;
    const textMatch=expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a,b)=>{
    if (sortBy==="date"){
      return a.createdAt < b.createdAt ? 1 : -1 ;
    } else if (sortBy==="amount") {
      return a.amount > b.amount ? 1 : -1 ;
    }

  });
}

//[1,4] ako uslov vrati -1 onda prvi broj dobija manji index i ide pre drugog i dobijamo rastucni niz
//a.amount > b.amount ? 1 : -1 ; za ovaj slucaj
//ako je zadati slucaj false prvi brooj ce da dobije manji index i gurace ga napred
//Ko je ranije napravljen neka ide prvi 
