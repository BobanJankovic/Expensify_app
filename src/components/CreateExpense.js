import React from 'react';
import { connect } from 'react-redux';
import AddForm from '../components/AddExpense';
import { startAddExpense } from '../actions/expenses';


 class CreateExpense extends React.Component {
    onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render(){
    return(
      <div>
        <AddForm onSubmit={this.onSubmit} />
      </div>
    )
  };
};

//da bih imao pristup dispatchu ubacujem dispatch to props ovako i koristim ga gore u nekoj funkciji.
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(CreateExpense);

/*
--------------------
Posto nemamo ovde mapStatetoProps stavljamo undefined
export default connect(mapStateToProps,mapDispatchToProps)(TodoItem)
--------------------
*/ 
/*
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}





*/