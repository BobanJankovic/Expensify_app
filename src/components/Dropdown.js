import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import {sortByAmount} from '../actions/filters';
import {sortByDate} from '../actions/filters';

const Dropdown= (props)=> (
  <div>
    <select value={props.filters.sortBy} 
      onChange={
                (e)=>{
                  if (e.target.value==="date") {
                    props.dispatch(sortByDate())
                  } else if ((e.target.value==="amount")){
                    props.dispatch(sortByAmount())
                  }
                }
              }
      >
      <option value="amount" >Amount</option>
      <option value="date" >Date</option>
      
    </select>
  </div>
);

const mapStateToProps = (state)=> {
  return {
    expenses:selectExpenses(state.expenses,state.filters),
    filters:state.filters
  };
};

export default connect(mapStateToProps)(Dropdown);
