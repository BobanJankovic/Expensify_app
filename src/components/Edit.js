import React from 'react';
import { connect } from 'react-redux';
import AddForm from './AddExpense';

import { startEditExpense,addExpense, removeExpense, startRemoveExpense } from '../actions/expenses';


const Edit = (props) => {
   console.log(props);
  return (
    <div>
    
      
      <AddForm
      //  This is edit od {props.expense.id}{props.expense.id} zbog linije koda 12 mi se vrednosti pojave u poljima koje treba da izmenim
      //props.expense pristupio sam iz stora i poslao je preko propsa
      expenseedit={props.expense}
       onSubmit={(expensed) => {
        
         props.dispatch(startEditExpense(props.expense.id, expensed));
          //props.dispatch(addExpense(expense));
          props.history.push('/');
         
        }}
         />
          <button  onClick={()=>{
            props.dispatch(startRemoveExpense({ id: props.expense.id}))
            props.history.push('/');
            }}>remove</button>
    </div>
  );
};


const mapStateToProps = (state, props) => {
  console.log(props);
  return {
    
    expense: state.expenses.find(expense => expense.id === props.match.params.id )
  };
};

export default connect(mapStateToProps)(Edit);
   
   /*import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
          props.dispatch(startEditExpense(props.expense.id, expense));
          props.history.push('/');
        }}
      />
      <button onClick={() => {
        props.dispatch(removeExpense({ id: props.expense.id }));
        props.history.push('/');
      }}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

export default connect(mapStateToProps)(EditExpensePage); */