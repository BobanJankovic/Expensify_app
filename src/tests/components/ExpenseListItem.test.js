import React from "react";
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {expenses} from '../fixtures/expenses';

Enzyme.configure({ adapter: new Adapter() });
//Ovo uvek kopiram kada testiram komponentu u ENZYME
//------------------------------------------------------------
//console.log(expenses[0])

test('should render ExpenseList correctly', () => {
const wrapper = shallow(<ExpenseListItem expenses={expenses} />);
expect(toJSON(wrapper)).toMatchSnapshot();
//expect(wrapper.find('h1').text()).toBe('Expensify');
//expect(renderer.getRenderOutput()).toMatchSnapshot();
//console.log(renderer.getRenderOutput());

}); 