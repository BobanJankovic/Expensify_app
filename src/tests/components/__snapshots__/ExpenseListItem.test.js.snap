// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList correctly 1`] = `
<div>
  <div>
    <div
      key="1"
    >
      <p>
        <Link
          replace={false}
          to="/edit/1"
        >
          Gum
        </Link>
        :
        195
         created at : 
        0
      </p>
    </div>
    <div
      key="2"
    >
      <p>
        <Link
          replace={false}
          to="/edit/2"
        >
          Rent
        </Link>
        :
        109500
         created at : 
        -345600000
      </p>
    </div>
    <div
      key="3"
    >
      <p>
        <Link
          replace={false}
          to="/edit/3"
        >
          Credit Card
        </Link>
        :
        4500
         created at : 
        345600000
      </p>
    </div>
  </div>
</div>
`;
